### DISCUSSION TITLE
Query on support for simple swizzles in AdaptiveCpp

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1294] [Q&A]
Hello,

I am currently working on a SYCL project where codebase needs to be compatible with both Intel's oneAPI and AdaptiveCpp. I have encountered an issue related to the use of simple swizzles, which are supported in oneAPI but seem to behave differently or are not supported in the same way in AdaptiveCpp.

When compiling code with oneAPI, I use the `#define SYCL_SIMPLE_SWIZZLES` preprocessor directive to enable simple swizzles. This works as expected. However, when compiling the same code with AdaptiveCpp, I run into errors. Specifically, calls to swizzles like `p.xyx()` result in compilation errors, indicating a possible lack of support or different implementation standards in AdaptiveCpp.

Here’s an example of the error I've encountering:

```
D:\source\repos\G2.cpp\Common\sycl\float_hash.h(42,46): error G0E51B380: no matching member function for call to 'xyx'
     42 |                 sycl::float3 p3 = fs::fract(sycl::float3(p.xyx()) * .1031f);
        |                                                          ~~^~~
```

I would like to understand if there is a way to enable simple swizzles in AdaptiveCpp that is compatible with how they are handled in oneAPI. 

Are simple swizzles supported in AdaptiveCpp? If so, could you provide guidance on how they can be used or enabled?

Thank you for your assistance!

**Comment by illuhad** [comment]
They can be used and enabled just like in DPC++:

For example, the following works:
```c++
#define SYCL_SIMPLE_SWIZZLES
#include <sycl/sycl.hpp>

int main()
{
  sycl::float3 p {1.f, 2.f, 3.f};
  sycl::float3 p3 = p.xyx();
  p3 *= .1031f;
  
  std::cout << p3.x() << " " << p3.y() << " " << p3.z() << "\n";
}
```

What is important is that the `SYCL_SIMPLE_SWIZZLES` macro is defined before the very first inclusion of `sycl.hpp`. If you have multiple files including `sycl.hpp`, the include that is actually the first one may not be the one you expect depending on the exact include order.

I cannot say anything more about what the problem might be based on the short snippet you have shown. Can you provide a full reproducer?

**Reply by blinkfrog** [reply]
Thank you, this is great to hear that simple swizzles are supported! 

I have done further testing and discovered that the issue is specifically related to using simple swizzles with float2 vectors. When attempting to use a swizzle like xyx on a float2, it results in a compilation error because the swizzle is constrained to vectors of size 3 or more ('2 >= 3' was not satisfied).

Here is a minimal example that reproduces the error:

```
#define SYCL_SIMPLE_SWIZZLES
#include <sycl/sycl.hpp>

int main(int argc, char* argv[]) {
    sycl::float2 p0 {1.0f, 2.0f};
    sycl::float3 p1 = p0.xyx(); // Error occurs here
    return 0;
}
```
The compilation error is:

```
C:/Temp/sycltest/SyclCmakeProjectSwizzle/swizzle.cpp:9:23: error: no matching member function for call to 'xyx'
    9 |         sycl::float3 p1 = p0.xyx();
      |                           ~~~^~~
C:\AdaptiveCpp\bin\..\include\AdaptiveCpp\sycl/../hipSYCL/sycl/libkernel/detail/vec_simple_swizzles.hpp:53:32: note: candidate template ignored: requirement '2 >= 3' was not satisfied [with n = 2]
   53 | HIPSYCL_DEFINE_VECTOR_SWIZZLE3(xyx, 3, 0, 1, 0)
      |                                ^
```
It appears that the AdaptiveCpp implementation follows the constraints very strictly, and thus using float2.xyx() is not supported. I believe this is the root cause of the problems I encountered in my larger project, as I was attempting to swizzle a float2 vector in this way.

I have been experimenting further to find a workaround for this float2 swizzle issue. I discovered that by adjusting the template constraint in the AdaptiveCpp implementation to allow a minimum of 2 components instead of 3, the code compiles and seems to swizzle correctly. Here's the change I made in the macro, defining template:

```
vec_simple_swizzles.hpp, line#53:
- HIPSYCL_DEFINE_VECTOR_SWIZZLE3(xyx, 3, 0, 1, 0)
+ HIPSYCL_DEFINE_VECTOR_SWIZZLE3(xyx, 2, 0, 1, 0)
```
With this change, the previously failing swizzle operation now compiles and behaves as expected in my initial tests.

Before considering this as a viable solution, I wanted to reach out and inquire about potential side effects. Could this change have any unforeseen consequences? If no, is it possible to make this change to AdaptiveCpp sources? Also, there are other swizzle templates that can be modified and enabled for 2-components vector swizzles, for example, `yxy` and others (and may be for 3-component-vectors too, need to analyze). If needed, I can create a PR.

**Reply by illuhad** [reply]
Thanks for looking into this! I've just checked the spec, and it says:

> Where XYZW_SWIZZLE is all permutations with repetition, of any subset with length greater than 1, of x, y for NumElements == 2, x, y, z for NumElements == 3 and x, y, z, w for NumElements == 4. For example a four element vec provides permutations including xzyw, xyyy and xz.

`NumElements` here is the the dimension of the `vec` on which `swizzle` is invoked. I'm not sure if `xyx` is legal here for a 2D `vec` because it is not a permutation of a subset of `x,y`.
Perhaps I am missing something. Do you read the spec differently? If DPC++ behaves differently, we should ask in the Khronos SYCL WG for clarification.

**Reply by blinkfrog** [reply]
Frankly, I haven't read the spec at all. Yes, DPC++ supports all possible swizzles for all vectors of dimension 2, 3 and 4, the same full set as glsl does.

At the moment, I've made required changes to both `vec_simple_swizzles.hpp` (updating `Dim` argument and adding it to 4D-vectors) and `vec.hpp` locally (updating `HIPSYCL_DEFINE_VECTOR_SWIZZLE4`), and all my codebase, using swizzles, compiles and works correctly. I can create PR, but, as you say that this is against the standard, then this probably would be a bad idea.

Edit: DPC++ even includes support of swizzles of 1D vectors:
```
    sycl::vec<float, 1> p0{ 1.0f };
    sycl::float4 p1 = p0.xxxx();
    std::cout << p1.x();
    std::cout << p1.y();
    std::cout << p1.z();
    std::cout << p1.w();
    ///output is "1111"
```
I added such support to my local version too.

**Reply by illuhad** [reply]
I have opened an issue in the SYCL specification repository: https://github.com/KhronosGroup/SYCL-Docs/issues/521
Seems like the spec is indeed unclear here, based on initial feedback, and needs to be clarified.
Let's wait for what the consensus there is.

**Reply by blinkfrog** [reply]
Thanks!

**Comment by blinkfrog** [comment]
I've written a python script which analyzes all current swizzle definitions and prints the vector dimension constraint changes that can be made to enable some of currently impossible swizzles. These are results:

```
{'xx': '2 -> 1', 'xxx': '3 -> 1', 'yxx': '3 -> 2', 'xyx': '3 -> 2', 'yyx': '3 -> 2', 'xxy': '3 -> 2', 'yxy': '3 -> 2', 'xyy': '3 -> 2', 'yyy': '3 -> 2'
```

However, 4-component swizzles always require 4-component vectors now because template doesn't take `Dim` parameter, so it isn't possible to take a float2 vector and swizzle it to, for example, `xyxy`.

**Comment by blinkfrog** [comment]
So, I propose to change template for SWIZZLE4 to this:
```
#define HIPSYCL_DEFINE_VECTOR_SWIZZLE4(name, Dim, i0, i1, i2, i3) \
  template<int n = N, \
           std::enable_if_t<(n >= Dim && n <= 4), bool> = true> \
  HIPSYCL_UNIVERSAL_TARGET \
  auto name() const \
  {return swizzle<i0,i1,i2,i3>(); }
  ```
  and add required dimensions as a first argument to swizzle macro definitions.
  
  These are required dimensions for SWIZZLE4:
  ```
  'xxxx': '1', 'rrrr': '1', 'yxxx': '1', 'grrr': '1', 'zxxx': '1', 'brrr': '1', 'wxxx': '1', 'arrr': '1', 'xyxx': '2', 'rgrr': '2', 'yyxx': '2', 'ggrr': '2', 'zyxx': '2', 'bgrr': '2', 'wyxx': '2', 'agrr': '2', 'xzxx': '3', 'rbrr': '3', 'yzxx': '3', 'gbrr': '3', 'zzxx': '3', 'bbrr': '3', 'wzxx': '3', 'abrr': '3', 'xwxx': '4', 'rarr': '4', 'ywxx': '4', 'garr': '4', 'zwxx': '4', 'barr': '4', 'wwxx': '4', 'aarr': '4', 'xxyx': '2', 'rrgr': '2', 'yxyx': '2', 'grgr': '2', 'zxyx': '2', 'brgr': '2', 'wxyx': '2', 'argr': '2', 'xyyx': '2', 'rggr': '2', 'yyyx': '2', 'gggr': '2', 'zyyx': '2', 'bggr': '2', 'wyyx': '2', 'aggr': '2', 'xzyx': '3', 'rbgr': '3', 'yzyx': '3', 'gbgr': '3', 'zzyx': '3', 'bbgr': '3', 'wzyx': '3', 'abgr': '3', 'xwyx': '4', 'ragr': '4', 'ywyx': '4', 'gagr': '4', 'zwyx': '4', 'bagr': '4', 'wwyx': '4', 'aagr': '4', 'xxzx': '3', 'rrbr': '3', 'yxzx': '3', 'grbr': '3', 'zxzx': '3', 'brbr': '3', 'wxzx': '3', 'arbr': '3', 'xyzx': '3', 'rgbr': '3', 'yyzx': '3', 'ggbr': '3', 'zyzx': '3', 'bgbr': '3', 'wyzx': '3', 'agbr': '3', 'xzzx': '3', 'rbbr': '3', 'yzzx': '3', 'gbbr': '3', 'zzzx': '3', 'bbbr': '3', 'wzzx': '3', 'abbr': '3', 'xwzx': '4', 'rabr': '4', 'ywzx': '4', 'gabr': '4', 'zwzx': '4', 'babr': '4', 'wwzx': '4', 'aabr': '4', 'xxwx': '4', 'rrar': '4', 'yxwx': '4', 'grar': '4', 'zxwx': '4', 'brar': '4', 'wxwx': '4', 'arar': '4', 'xywx': '4', 'rgar': '4', 'yywx': '4', 'ggar': '4', 'zywx': '4', 'bgar': '4', 'wywx': '4', 'agar': '4', 'xzwx': '4', 'rbar': '4', 'yzwx': '4', 'gbar': '4', 'zzwx': '4', 'bbar': '4', 'wzwx': '4', 'abar': '4', 'xwwx': '4', 'raar': '4', 'ywwx': '4', 'gaar': '4', 'zwwx': '4', 'baar': '4', 'wwwx': '4', 'aaar': '4', 'xxxy': '2', 'rrrg': '2', 'yxxy': '2', 'grrg': '2', 'zxxy': '2', 'brrg': '2', 'wxxy': '2', 'arrg': '2', 'xyxy': '2', 'rgrg': '2', 'yyxy': '2', 'ggrg': '2', 'zyxy': '2', 'bgrg': '2', 'wyxy': '2', 'agrg': '2', 'xzxy': '3', 'rbrg': '3', 'yzxy': '3', 'gbrg': '3', 'zzxy': '3', 'bbrg': '3', 'wzxy': '3', 'abrg': '3', 'xwxy': '4', 'rarg': '4', 'ywxy': '4', 'garg': '4', 'zwxy': '4', 'barg': '4', 'wwxy': '4', 'aarg': '4', 'xxyy': '2', 'rrgg': '2', 'yxyy': '2', 'grgg': '2', 'zxyy': '2', 'brgg': '2', 'wxyy': '2', 'argg': '2', 'xyyy': '2', 'rggg': '2', 'yyyy': '2', 'gggg': '2', 'zyyy': '2', 'bggg': '2', 'wyyy': '2', 'aggg': '2', 'xzyy': '3', 'rbgg': '3', 'yzyy': '3', 'gbgg': '3', 'zzyy': '3', 'bbgg': '3', 'wzyy': '3', 'abgg': '3', 'xwyy': '4', 'ragg': '4', 'ywyy': '4', 'gagg': '4', 'zwyy': '4', 'bagg': '4', 'wwyy': '4', 'aagg': '4', 'xxzy': '3', 'rrbg': '3', 'yxzy': '3', 'grbg': '3', 'zxzy': '3', 'brbg': '3', 'wxzy': '3', 'arbg': '3', 'xyzy': '3', 'rgbg': '3', 'yyzy': '3', 'ggbg': '3', 'zyzy': '3', 'bgbg': '3', 'wyzy': '3', 'agbg': '3', 'xzzy': '3', 'rbbg': '3', 'yzzy': '3', 'gbbg': '3', 'zzzy': '3', 'bbbg': '3', 'wzzy': '3', 'abbg': '3', 'xwzy': '4', 'rabg': '4', 'ywzy': '4', 'gabg': '4', 'zwzy': '4', 'babg': '4', 'wwzy': '4', 'aabg': '4', 'xxwy': '4', 'rrag': '4', 'yxwy': '4', 'grag': '4', 'zxwy': '4', 'brag': '4', 'wxwy': '4', 'arag': '4', 'xywy': '4', 'rgag': '4', 'yywy': '4', 'ggag': '4', 'zywy': '4', 'bgag': '4', 'wywy': '4', 'agag': '4', 'xzwy': '4', 'rbag': '4', 'yzwy': '4', 'gbag': '4', 'zzwy': '4', 'bbag': '4', 'wzwy': '4', 'abag': '4', 'xwwy': '4', 'raag': '4', 'ywwy': '4', 'gaag': '4', 'zwwy': '4', 'baag': '4', 'wwwy': '4', 'aaag': '4', 'xxxz': '3', 'rrrb': '3', 'yxxz': '3', 'grrb': '3', 'zxxz': '3', 'brrb': '3', 'wxxz': '3', 'arrb': '3', 'xyxz': '3', 'rgrb': '3', 'yyxz': '3', 'ggrb': '3', 'zyxz': '3', 'bgrb': '3', 'wyxz': '3', 'agrb': '3', 'xzxz': '3', 'rbrb': '3', 'yzxz': '3', 'gbrb': '3', 'zzxz': '3', 'bbrb': '3', 'wzxz': '3', 'abrb': '3', 'xwxz': '4', 'rarb': '4', 'ywxz': '4', 'garb': '4', 'zwxz': '4', 'barb': '4', 'wwxz': '4', 'aarb': '4', 'xxyz': '3', 'rrgb': '3', 'yxyz': '3', 'grgb': '3', 'zxyz': '3', 'brgb': '3', 'wxyz': '3', 'argb': '3', 'xyyz': '3', 'rggb': '3', 'yyyz': '3', 'gggb': '3', 'zyyz': '3', 'bggb': '3', 'wyyz': '3', 'aggb': '3', 'xzyz': '3', 'rbgb': '3', 'yzyz': '3', 'gbgb': '3', 'zzyz': '3', 'bbgb': '3', 'wzyz': '3', 'abgb': '3', 'xwyz': '4', 'ragb': '4', 'ywyz': '4', 'gagb': '4', 'zwyz': '4', 'bagb': '4', 'wwyz': '4', 'aagb': '4', 'xxzz': '3', 'rrbb': '3', 'yxzz': '3', 'grbb': '3', 'zxzz': '3', 'brbb': '3', 'wxzz': '3', 'arbb': '3', 'xyzz': '3', 'rgbb': '3', 'yyzz': '3', 'ggbb': '3', 'zyzz': '3', 'bgbb': '3', 'wyzz': '3', 'agbb': '3', 'xzzz': '3', 'rbbb': '3', 'yzzz': '3', 'gbbb': '3', 'zzzz': '3', 'bbbb': '3', 'wzzz': '3', 'abbb': '3', 'xwzz': '4', 'rabb': '4', 'ywzz': '4', 'gabb': '4', 'zwzz': '4', 'babb': '4', 'wwzz': '4', 'aabb': '4', 'xxwz': '4', 'rrab': '4', 'yxwz': '4', 'grab': '4', 'zxwz': '4', 'brab': '4', 'wxwz': '4', 'arab': '4', 'xywz': '4', 'rgab': '4', 'yywz': '4', 'ggab': '4', 'zywz': '4', 'bgab': '4', 'wywz': '4', 'agab': '4', 'xzwz': '4', 'rbab': '4', 'yzwz': '4', 'gbab': '4', 'zzwz': '4', 'bbab': '4', 'wzwz': '4', 'abab': '4', 'xwwz': '4', 'raab': '4', 'ywwz': '4', 'gaab': '4', 'zwwz': '4', 'baab': '4', 'wwwz': '4', 'aaab': '4', 'xxxw': '4', 'rrra': '4', 'yxxw': '4', 'grra': '4', 'zxxw': '4', 'brra': '4', 'wxxw': '4', 'arra': '4', 'xyxw': '4', 'rgra': '4', 'yyxw': '4', 'ggra': '4', 'zyxw': '4', 'bgra': '4', 'wyxw': '4', 'agra': '4', 'xzxw': '4', 'rbra': '4', 'yzxw': '4', 'gbra': '4', 'zzxw': '4', 'bbra': '4', 'wzxw': '4', 'abra': '4', 'xwxw': '4', 'rara': '4', 'ywxw': '4', 'gara': '4', 'zwxw': '4', 'bara': '4', 'wwxw': '4', 'aara': '4', 'xxyw': '4', 'rrga': '4', 'yxyw': '4', 'grga': '4', 'zxyw': '4', 'brga': '4', 'wxyw': '4', 'arga': '4', 'xyyw': '4', 'rgga': '4', 'yyyw': '4', 'ggga': '4', 'zyyw': '4', 'bgga': '4', 'wyyw': '4', 'agga': '4', 'xzyw': '4', 'rbga': '4', 'yzyw': '4', 'gbga': '4', 'zzyw': '4', 'bbga': '4', 'wzyw': '4', 'abga': '4', 'xwyw': '4', 'raga': '4', 'ywyw': '4', 'gaga': '4', 'zwyw': '4', 'baga': '4', 'wwyw': '4', 'aaga': '4', 'xxzw': '4', 'rrba': '4', 'yxzw': '4', 'grba': '4', 'zxzw': '4', 'brba': '4', 'wxzw': '4', 'arba': '4', 'xyzw': '4', 'rgba': '4', 'yyzw': '4', 'ggba': '4', 'zyzw': '4', 'bgba': '4', 'wyzw': '4', 'agba': '4', 'xzzw': '4', 'rbba': '4', 'yzzw': '4', 'gbba': '4', 'zzzw': '4', 'bbba': '4', 'wzzw': '4', 'abba': '4', 'xwzw': '4', 'raba': '4', 'ywzw': '4', 'gaba': '4', 'zwzw': '4', 'baba': '4', 'wwzw': '4', 'aaba': '4', 'xxww': '4', 'rraa': '4', 'yxww': '4', 'graa': '4', 'zxww': '4', 'braa': '4', 'wxww': '4', 'araa': '4', 'xyww': '4', 'rgaa': '4', 'yyww': '4', 'ggaa': '4', 'zyww': '4', 'bgaa': '4', 'wyww': '4', 'agaa': '4', 'xzww': '4', 'rbaa': '4', 'yzww': '4', 'gbaa': '4', 'zzww': '4', 'bbaa': '4', 'wzww': '4', 'abaa': '4', 'xwww': '4', 'raaa': '4', 'ywww': '4', 'gaaa': '4', 'zwww': '4', 'baaa': '4', 'wwww': '4', 'aaaa': '4'}
  ```
